### PR TITLE
Dont allow directly dialing to servers not in inventory

### PR DESCRIPTION
**Pull request by lxea**
Fix for #28984 

This feature is being deprecated and so does not require adding a switch to re-enable this, just that it get removed.

### CODE REVIEW

**Review by GavinFrazar — APPROVED** [review]

**Comment by codingllama** [comment]
I think you got some legitimate test breakages?

I don't have much background in the decision, couldn't this cause problems to existing installs? Should we ask @zmb3 or @r0mant to approve?

**Comment by lxea** [comment]
> I think you got some legitimate test breakages?
> 
I forgot to push it earlier, I have a fix for the unit tests, I'm still figuring out the fix for the integration tests, its failing as it makes use of a node not in the inventory for testing host key checking

File: lib/proxy/router.go [hunk 1]
```diff
@@ -260,12 +258,7 @@ func (r *Router) DialHost(ctx context.Context, clientSrcAddr, clientDstAddr net.
 			serverAddr = reversetunnelclient.LocalNode
 		}
 	} else {
-		if port == "" || port == "0" {
-			port = strconv.Itoa(defaults.SSHServerListenPort)
-		}
-
-		serverAddr = net.JoinHostPort(host, port)
-		r.log.Warnf("server lookup failed: using default=%v", serverAddr)
+		return nil, trace.NotImplemented("direct dialing to nodes not found in inventory is not supported")
```
Comment (zmb3) [inline lib/proxy/router.go:261]: Can we invert this check and get rid of the else?

```
if target == nil {
    return nil, trace.NotImplemented("...")
}
isAgentLessNode = target.GetSubKind() == types.SubKindOpenSSHNode
...
```

**Comment by codingllama** [comment]
Thanks for the explanation, folks.

@lxea, let us know when you are ready for another look.

**Comment by klizhentas** [comment]
To add some context:

The existing box bypasses RBAC. We want to get rid of this functionality to remove the RBAC bypass.

Now any user want to access SSH nodes, all they need to do is to register the target node in the registry

So  although it's a breaking change, it's worth it because it improves security and removes RBAC exception from the system.

**Comment by fspmarshall** [comment]
Due to the potentially significant impact this change might have on a user that hasn't been following changelog/documentation closely, we need to provide an escape hatch to ensure that users can regain access to their agents if they need to perform agent-side actions in order to go through the new enroll process.

I've thrown together a proposal that I think should work (based off of this branch, so you should be able to just cherry-pick it in if it works for you):

-  https://github.com/gravitational/teleport/commit/7eb1d4656301c561ac0531088fe6dfe9df806b09

The above commit causes the proxy to revert to the old behavior when `TELEPORT_UNSTABLE_UNLISTED_AGENT_DIALING=yes` is set in its env.

File: lib/proxy/router.go [hunk 2]
```diff
@@ -160,13 +162,14 @@ func (c *RouterConfig) CheckAndSetDefaults() error {
...
-	tracer         oteltrace.Tracer
-	serverResolver serverResolverFn
+	clusterName           string
+	log                   *logrus.Entry
+	clusterGetter         RemoteClusterGetter
+	localSite             reversetunnelclient.RemoteSite
+	siteGetter            SiteGetter
+	tracer                oteltrace.Tracer
+	serverResolver        serverResolverFn
+	permitUnlistedDialing bool
```
Comment (fspmarshall) [inline lib/proxy/router.go:172]: Spaced it before, but we probably want a DELETE IN directive for this.

```suggestion
	// DELETE IN 15.0.0: necessary for smoothing over v13 to v14 transition only.
	permitUnlistedDialing bool
```

**Review by fspmarshall — APPROVED** [review]

**Comment by zmb3** [comment]
Do we understand why the tests broke after this change? Just running them for 30 seconds and hoping they eventually succeed without understanding what changed worries me.

**Comment by lxea** [comment]
> Do we understand why the tests broke after this change? Just running them for 30 seconds and hoping they eventually succeed without understanding what changed worries me.

AFAIU the tests were breaking as there seems to be a race between the ssh server starting and actually being in the inventory and this didnt matter before

**Comment by lxea** [comment]
> > AFAIU the tests were breaking as there seems to be a race between the ssh server starting and actually being in the inventory and this didnt matter before
> 
> Okay, if that's the case then we should add a `require.Eventually` that waits for the node to show up in inventory. What you have here is treating the symptom, not the cause.

I've changed to this.

I've also skipped `TestSSHLoadAllCAs` in the flaky test detector, TestProxySSH is already skipped as it takes a long time to bring up the tsh suite and `TestSSHLoadAllCAs` is slow for the same reason

**Pull request merged**